Exclude nodes that should be skipped because of directives in getChildren

xuorig opened the pull request:
Fixes #536 

I've been working with graphql-ruby and relay and been wanting to use directives. I found #536 yesterday night and tried to work on a solution today.

Following what @josephsavona said in the issue, I've modified`RelayQueryNode.getChildren` so it only return children that are not skipped (meaning `skip if: true` and `include if: false`). I've implemented the function `RelayQueryNode.shouldBeIncluded` which
  - Checks if a skip directive is present ( It has priority over include)
     - If an if argument is present, return the negated value
  - Checks if an include directive is present
     - if an if argument, return the value
  - return true in all other cases

I've tried to take into consideration that other directives and arguments other than `if:` might be added, which might make the code a little bit more complicated, let me know if I should just assume that the argument is if right now.

Let me know if the solution makes sense. I've added 2 tests, modified a printing test that included a field that should be skipped and also modified the metric test since the getChildren method now calls getDirectives for every field.

Haven't had much chance to look at Relay in detail so let me know how I can improve this!

josephsavona commented on src/query/RelayQuery.js:
```diff
@@ -214,6 +218,36 @@ class RelayQueryNode {
+      }
+    }
+
+    return true
```
Because the schema may change, it's definitely good to scan the arguments for each directive. To make this more performant, how about iterating over the directives only once? Something like

```javascript
let isSkipped = false;
let isIncluded = true;
directives.forEach(directive => {
  if (directive.name === SKIP) {
    isSkipped = ...
  } else if (directive.name === INCLUDE) {
    isIncluded = ...
  }
});
return isSkipped || !isIncluded;
```

xuorig replied:
Good idea for the iteration, I will add that in tonight. I dont think `isSkipped || !isIncluded` will work though. Since `skip` should always have priority over `include`, something like `False || !False` would result in the field being skipped even though skip was `False`!

josephsavona commented on src/query/RelayQuery.js:
```diff
@@ -214,6 +218,36 @@ class RelayQueryNode {
     return children;
   }
 
+  shouldBeIncluded(): ?boolean {
```
:+1: for a new method.

How about naming it `isSkipped(): boolean` to match the `isScalar()`, `isGenerated` pattern? Also note that the return type can be a non-null `boolean`.

josephsavona commented on src/query/__tests__/RelayQueryRoot-test.js:
```diff
@@ -73,6 +73,38 @@ describe('RelayQueryRoot', () => {
+        me {
+          id,
+          firstName @skip(if: $cond),
+          lastName @include(if: $cond2),
```
nit: variable names are arbitrary, so you could name this `@skip(if: $true)` to make it even more clear that the variable's value will be `true`

josephsavona commented:
@xuorig This is awesome!!! I have some minor feedback but overall looks good.

josephsavona commented:
@facebook-github-bot import

xuorig commented:
@josephsavona I've addressed your comments. Let me know what you think about my solution for iterating only once over directives. Not the most pretty but it's hard to make it with that Skip > Include priority.
